env/env_encrypt2.cc

```diff
@@ -0,0 +1,989 @@
+
+  Slice marker_slice(marker, sizeof(EncryptMarker));
+  status = f->Write(0, marker_slice);
+
+  if (status.ok()) {
+    // create nonce, then write it and key description
+    Slice prefix_slice((char*)&prefix, sizeof(prefix));
+
+    status = provider->CreateNewPrefix(std::string(), (char*)&prefix,
+                                       provider->GetPrefixLength());
+
+    if (status.ok()) {
+      status = f->Write(sizeof(EncryptMarker), prefix_slice);
+    }
+  }
+
+  if (status.ok()) {
+    stream.reset(new AESBlockAccessCipherStream(provider->key(), code_version,
+                                                prefix.nonce_));
+  }
+
+  return status;
+}
+
+// NewSequentialFile opens a file for sequential reading.
+Status EncryptedEnvV2::NewSequentialFile(
+    const std::string& fname, std::unique_ptr<SequentialFile>* result,
+    const EnvOptions& options) {
+  result->reset();
+  if (options.use_mmap_reads || options.use_direct_reads) {
+    return Status::InvalidArgument();
+  }
+
+  // Open file using underlying Env implementation
+  std::unique_ptr<SequentialFile> underlying;
+  auto status = EnvWrapper::NewSequentialFile(fname, &underlying, options);
+  if (status.ok()) {
+    std::shared_ptr<const CTREncryptionProviderV2> provider;
+    std::unique_ptr<BlockAccessCipherStream> stream;
+    status = ReadSeqEncryptionPrefix<SequentialFile>(underlying.get(), provider,
+                                                     stream);
+
+    if (status.ok()) {
+      if (provider) {
+        (*result) = std::unique_ptr<SequentialFile>(new EncryptedSequentialFile(
+            std::move(underlying), std::move(stream),
+            provider->GetPrefixLength()));
+
+      } else {
+        // normal file, not encrypted
+        // sequential file might not allow backing up to begining, close and
+        // reopen
+        underlying.reset(nullptr);
+        status = EnvWrapper::NewSequentialFile(fname, result, options);
+      }
+    }
+  }
+
+  return status;
+}
+
+// NewRandomAccessFile opens a file for random read access.
+Status EncryptedEnvV2::NewRandomAccessFile(
+    const std::string& fname, std::unique_ptr<RandomAccessFile>* result,
+    const EnvOptions& options) {
+  result->reset();
+  if (options.use_mmap_reads || options.use_direct_reads) {
+    return Status::InvalidArgument();
+  }
+
+  // Open file using underlying Env implementation
+  std::unique_ptr<RandomAccessFile> underlying;
+  auto status = EnvWrapper::NewRandomAccessFile(fname, &underlying, options);
+  if (status.ok()) {
+    std::shared_ptr<const CTREncryptionProviderV2> provider;
+    std::unique_ptr<BlockAccessCipherStream> stream;
+    status = ReadRandEncryptionPrefix<RandomAccessFile>(underlying.get(),
+                                                        provider, stream);
+
+    if (status.ok()) {
+      if (provider) {
+        (*result) =
+            std::unique_ptr<RandomAccessFile>(new EncryptedRandomAccessFile(
```

Comment:
Shouldn't this be RandomAccessFile2?

Reply:
There is no EncryptedRandomAccessFileV2.  EncryptedRandomAccessFile class could be used as is.  Only EncryptedWriteableFile and EncryptedRandRWFile needed adjustment.